tests/test_schema_retrieval.py

```diff
@@ -26,29 +26,34 @@
 class TestSchemaRetrieval:
     "Test schema retrieval."
 
+    # We allow this to fail because the default remote might be down but it's not this library's issue.
+    @pytest.mark.xfail
     def test_default_schema(self):
         "Test the basic functioning of retrieving default schema files."
 
-        # We assert they are identical to the test schema files for now, because we don't host them on any websites.
-        # Probably this will change in the future.
-        assert retrieve_schema_file(DatasetSchema.DEFAULT_SCHEMA_URL) == \
-            Path('tests/schemata/datasets.yaml').read_text()
-        assert retrieve_schema_file(FormatSchema.DEFAULT_SCHEMA_URL) == \
-            Path('tests/schemata/formats.yaml').read_text()
-        assert retrieve_schema_file(LicenseSchema.DEFAULT_SCHEMA_URL) == \
-            Path('tests/schemata/licenses.yaml').read_text()
+        # We only assert that we have retrieved some non-empty files here. This is because we want to decouple the
+        # maintenance of schema files in production with the library development. These files likely would change more
+        # regularly than the library.
+        assert retrieve_schema_file(DatasetSchema.DEFAULT_SCHEMA_URL)
+        assert retrieve_schema_file(FormatSchema.DEFAULT_SCHEMA_URL)
+        assert retrieve_schema_file(LicenseSchema.DEFAULT_SCHEMA_URL)
 
-    def test_custom_schema(self):
-        "Test retrieving user-specified schema files."
+    def test_custom_schema_local(self):
+        "Test retrieving user-specified local schema files."
 
         # We assert they are identical to the test schema files for now, because we don't host them on any websites.
         # Probably this will change in the future.
         assert retrieve_schema_file('./tests/schemata/datasets.yaml') == \
             Path('tests/schemata/datasets.yaml').read_text()
         assert retrieve_schema_file(f'file://{os.getcwd()}/tests/schemata/formats.yaml') == \
             Path('tests/schemata/formats.yaml').read_text()
-        assert retrieve_schema_file(LicenseSchema.DEFAULT_SCHEMA_URL.replace('https://', 'http://', 1)) == \
+
+    def test_custom_schema_remote(self, http_schema_file_url):
+        "Test retrieving user-specified remote schema files."
+
+        assert retrieve_schema_file(http_schema_file_url + 'licenses.yaml') == \
             Path('tests/schemata/licenses.yaml').read_text()
+        # TODO: Add https tests here when we add stronger security measurements
```

Comment:
Just so I check my understanding: this is mimicking remotely retrieving a custom schema by setting up an HTTP server that serves files from the current directory. So we basically spoof `licenses.yml` as if it's hosted on a website and then feed `retrieve_schema_file` the spoofed url, right?

If the HTTP server we created is blocking does that mean without a separate thread, it stops some part of the python script from running until `retrieve_schema_file(http_schema_file_url + 'licenses.yaml')` executes?

Do we need to clean up the thread like we did with the symlink in `tmp_symlink_dir`?

Reply:
> Just so I check my understanding: this is mimicking remotely retrieving a custom schema by setting up an HTTP server that serves files from the current directory. So we basically spoof `licenses.yml` as if it's hosted on a website and then feed `retrieve_schema_file` the spoofed url, right?

Yes.
> 
> If the HTTP server we created is blocking does that mean without a separate thread, it stops some part of the python script from running until `retrieve_schema_file(http_schema_file_url + 'licenses.yaml')` executes?

Are you referring to the threading part? If we don't have the threading part, the test would stop at `serve_forever` and nothing more will be executed (because we are running in a single-threaded fashion).

> 
> Do we need to clean up the thread like we did with the symlink in `tmp_symlink_dir`?

It's cleaned up, because the http server starts serving in a context manager. Once it exits the context manager, `serve_forever` exists and the http thread terminates. If the test terminates early for other reasons, the thread is set to be daemonic so that it will also exit.

Reply:
got it thanks for the explanation!